E00062 could be more specific about which command line option can't be parsed

E0062 is unclear, we currently get `error[E0062]: unknown value for option. Valid values are 'none, source-map, default, standard, glsl'` when we could have `error[E0062]: unknown value for option -line-directive-mode. Valid values are 'none, source-map, default, standard, glsl'`

--- a/source/slang/slang-diagnostics.lua
+++ b/source/slang/slang-diagnostics.lua
@@ -306,7 +306,11 @@ err(
     "no specified '-target' option matches the output path '~path', which implies the '~format' format"
 )
 
-err("unknown-command-line-value", 62, "unknown value for option. Valid values are '~validValues'")
+err(
+    "unknown-command-line-value",
+    62,
+    "unknown value for option '~option'. Valid values are '~validValues'"
+)
 
 err("unknown-help-category", 63, "unknown help category")
 

--- a/source/slang/slang-options.cpp
+++ b/source/slang/slang-options.cpp
@@ -1384,6 +1384,10 @@ struct OptionsParser
 
     CommandOptions* m_cmdOptions = nullptr;
     CommandLineContext* m_cmdLineContext = nullptr;
+
+    // Tracks the currently-being-parsed option name (e.g. "-line-directive-mode")
+    // so that diagnostics like UnknownCommandLineValue can mention it.
+    String m_currentOptionName;
 };
 
 int OptionsParser::addTranslationUnit(SlangSourceLanguage language, Stage impliedStage)
@@ -1862,7 +1866,9 @@ SlangResult OptionsParser::_getValue(
         StringBuilder buf;
         StringUtil::join(names.getBuffer(), names.getCount(), toSlice(", "), buf);
 
-        m_sink->diagnose(Diagnostics::UnknownCommandLineValue{.validValues = buf});
+        m_sink->diagnose(Diagnostics::UnknownCommandLineValue{
+            .option = m_currentOptionName,
+            .validValues = buf});
         return SLANG_FAIL;
     }
 
@@ -1916,7 +1922,8 @@ SlangResult OptionsParser::_getValue(
     StringBuilder buf;
     StringUtil::join(names.getBuffer(), names.getCount(), toSlice(", "), buf);
 
-    m_sink->diagnose(Diagnostics::UnknownCommandLineValue{.validValues = buf});
+    m_sink->diagnose(
+        Diagnostics::UnknownCommandLineValue{.option = m_currentOptionName, .validValues = buf});
     return SLANG_FAIL;
 }
 
@@ -2366,6 +2373,12 @@ SlangResult OptionsParser::_parse(int argc, char const* const* argv)
             return SLANG_FAIL;
         }
 
+        // Use the canonical first-listed name for the option rather than
+        // the raw argument token. For prefix-style options like `-O3` /
+        // `-g2`, `argValue` includes the suffix; the canonical name
+        // (e.g. `-O`, `-g`) is what the user is supposed to recognise.
+        m_currentOptionName = m_cmdOptions->getFirstNameForOption(optionIndex);
+
         const auto optionKind = OptionKind(m_cmdOptions->getOptionAt(optionIndex).userValue);
 
         switch (optionKind)
@@ -2453,8 +2466,9 @@ SlangResult OptionsParser::_parse(int argc, char const* const* argv)
                     colorValue = SLANG_DIAGNOSTIC_COLOR_AUTO;
                 else
                 {
-                    m_sink->diagnose(
-                        Diagnostics::UnknownCommandLineValue{.validValues = "always, never, auto"});
+                    m_sink->diagnose(Diagnostics::UnknownCommandLineValue{
+                        .option = m_currentOptionName,
+                        .validValues = "always, never, auto"});
                     return SLANG_FAIL;
                 }
                 linkage->m_optionSet.set(optionKind, (int)colorValue);

--- a/tests/diagnostics/command-line/unknown-debug-info-level.slang
+++ b/tests/diagnostics/command-line/unknown-debug-info-level.slang
@@ -1,0 +1,9 @@
+// unknown-debug-info-level.slang
+//
+// Coverage for the multi-category `_getValue` path: `-g...` accepts
+// values from two `ValueCategory`s (DebugLevel + DebugInfoFormat) and
+// joins their names for the diagnostic. Pin the canonical option name
+// here as well.
+
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):-gbogus
+//CHECK: unknown value for option '-g'. Valid values are '0, none, 1, minimal, 2, standard, 3, maximal, default-format, c7, pdb, stabs, coff, dwarf'

--- a/tests/diagnostics/command-line/unknown-diagnostic-color.slang
+++ b/tests/diagnostics/command-line/unknown-diagnostic-color.slang
@@ -1,0 +1,4 @@
+// unknown-diagnostic-color.slang
+
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):-diagnostic-color magenta
+//CHECK: unknown value for option '-diagnostic-color'. Valid values are 'always, never, auto'

--- a/tests/diagnostics/command-line/unknown-line-directive-mode.slang
+++ b/tests/diagnostics/command-line/unknown-line-directive-mode.slang
@@ -1,4 +1,4 @@
 // unknown-line-directive-mode.slang
 
 //DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):-line-directive-mode quizzical
-//CHECK: unknown value for option. Valid values are 'none, source-map, default, standard, glsl'
+//CHECK: unknown value for option '-line-directive-mode'. Valid values are 'none, source-map, default, standard, glsl'

--- a/tests/diagnostics/command-line/unknown-optimization-level.slang
+++ b/tests/diagnostics/command-line/unknown-optimization-level.slang
@@ -1,0 +1,7 @@
+// unknown-optimization-level.slang
+//
+// Regression coverage for the prefix-style option case: `-O...` should
+// report the canonical `-O` rather than the full token `-Obogus`.
+
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):-Obogus
+//CHECK: unknown value for option '-O'. Valid values are '0, none, 1, default, 2, high, 3, maximal'

--- a/tests/diagnostics/execution-model/mesh-output-array-no-size.slang
+++ b/tests/diagnostics/execution-model/mesh-output-array-no-size.slang
@@ -10,6 +10,10 @@ struct VertexOutput
 [outputtopology("triangle")]
 [numthreads(1, 1, 1)]
 void meshMain(
+/*CHECK:
+     ^^^^^^^^ mesh shader missing OutputVertices/OutputIndices output
+     ^^^^^^^^ mesh shader entry point 'meshMain' must declare both an 'OutputVertices' and an 'OutputIndices' output parameter
+*/
     out vertices VertexOutput verts[],
     /*CHECK:
                               ^^^^^ mesh output array must have size

--- a/tests/diagnostics/execution-model/mesh-output-not-array.slang
+++ b/tests/diagnostics/execution-model/mesh-output-not-array.slang
@@ -10,6 +10,10 @@ struct VertexOutput
 [outputtopology("triangle")]
 [numthreads(1, 1, 1)]
 void meshMain(
+/*CHECK:
+     ^^^^^^^^ mesh shader missing OutputVertices/OutputIndices output
+     ^^^^^^^^ mesh shader entry point 'meshMain' must declare both an 'OutputVertices' and an 'OutputIndices' output parameter
+*/
     out vertices VertexOutput verts,
     /*CHECK:
                               ^^^^^ mesh output must be array

